File: Module5/Case/RestFurama/src/main/java/thanh/code/models/Employee.java
package thanh.code.models;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;

import javax.annotation.Resource;
import javax.persistence.*;
import javax.validation.Valid;
import javax.validation.constraints.Email;
import javax.validation.constraints.NotBlank;
import java.sql.Date;
import java.util.Set;

@Entity
public class Employee {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private int employeeId;
    @NotBlank
    private String employeeName;
    private Date employeeBirthday;
    private String employeeIdCard;
    private Double employeeSalary;
    private String employeePhone;
    @Email
    private String employeeEmail;
    private String employeeAddress;

    @ManyToOne(targetEntity = Position.class)
    @Resource
    @JsonBackReference(value = "position")
    private Position position;

    @ManyToOne(targetEntity = EducationDegree.class)
    @Resource
    @JsonBackReference(value = "educationDegree")
    private EducationDegree educationDegree;

    @ManyToOne(targetEntity = Division.class)
    @Resource
    @JsonBackReference(value = "division")
    private Division division;

    @Valid
    @OneToOne(targetEntity = User.class, cascade = CascadeType.ALL)
    @Resource
    @JsonBackReference(value = "user")
    private User user;

    @OneToMany(mappedBy = "employee", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JsonManagedReference(value = "employee")
    private Set<Contract> contract;

    public Set<Contract> getContract() {
        return contract;
    }

    public void setContract(Set<Contract> contract) {
        this.contract = contract;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public Division getDivision() {
        return division;
    }

    public void setDivision(Division division) {
        this.division = division;
    }

    public Position getPosition() {
        return position;
    }

    public void setPosition(Position position) {
        this.position = position;
    }

    public EducationDegree getEducationDegree() {
        return educationDegree;
    }

    public void setEducationDegree(EducationDegree educationDegree) {
        this.educationDegree = educationDegree;
    }

    public int getEmployeeId() {
        return employeeId;
    }

    public void setEmployeeId(int employeeId) {
        this.employeeId = employeeId;
    }

    public String getEmployeeName() {
        return employeeName;
    }

    public void setEmployeeName(String employeeName) {
        this.employeeName = employeeName;
    }

    public Date getEmployeeBirthday() {
        return employeeBirthday;
    }

    public void setEmployeeBirthday(Date employeeBirthday) {
        this.employeeBirthday = employeeBirthday;
    }

    public String getEmployeeIdCard() {
        return employeeIdCard;
    }

    public void setEmployeeIdCard(String employeeIdCard) {
        this.employeeIdCard = employeeIdCard;
    }

    public Double getEmployeeSalary() {
        return employeeSalary;
    }

    public void setEmployeeSalary(Double employeeSalary) {
        this.employeeSalary = employeeSalary;
    }

    public String getEmployeePhone() {
        return employeePhone;
    }

    public void setEmployeePhone(String employeePhone) {
        this.employeePhone = employeePhone;
    }

    public String getEmployeeEmail() {
        return employeeEmail;
    }

    public void setEmployeeEmail(String employeeEmail) {
        this.employeeEmail = employeeEmail;
    }

    public String getEmployeeAddress() {
        return employeeAddress;
    }

    public void setEmployeeAddress(String employeeAddress) {
        this.employeeAddress = employeeAddress;
    }

    public Employee() {
    }
}

File: Module5/Case/Furama/src/app/models/employee.ts
import {Position} from "./position";
import {IEducationDegree} from "./education-degree";
import {IDivision} from "./division";
import {IUser} from "./user";

export class Employee implements IEmployee{
  division: IDivision;
  educationDegree: IEducationDegree;
  employeeAddress: string;
  employeeBirthday: string;
  employeeEmail: string;
  employeeId: number;
  employeeIdCard: string;
  employeeName: string;
  employeePhone: string;
  employeeSalary: number;
  position: Position;
  user: IUser;

  constructor(division: IDivision, educationDegree: IEducationDegree, employeeAddress: string, employeeBirthday: string,
              employeeEmail: string, employeeId: number, employeeIdCard: string, employeeName: string,
              employeePhone: string, employeeSalary: number, position: Position, userName: IUser) {
    this.division = division;
    this.educationDegree = educationDegree;
    this.employeeAddress = employeeAddress;
    this.employeeBirthday = employeeBirthday;
    this.employeeEmail = employeeEmail;
    this.employeeId = employeeId;
    this.employeeIdCard = employeeIdCard;
    this.employeeName = employeeName;
    this.employeePhone = employeePhone;
    this.employeeSalary = employeeSalary;
    this.position = position;
    this.user = userName;
  }
}

export interface IEmployee {
  employeeId: number;
  employeeName: string;
  employeeBirthday: string;
  employeeIdCard: string;
  employeeSalary: number;
  employeePhone: string;
  employeeEmail: string;
  employeeAddress: string;

  position: Position;
  educationDegree: IEducationDegree;
  division: IDivision;
  user: IUser;
}
